### ISSUE TITLE
Option to remove "=" sign from `ModelCheckpoint` ckpt file names

### ISSUE THREAD

**Issue by Anjum48** [Lightning-AI/lightning#17825]
### Description & Motivation

Currently `ModelCheckpoint` saves checkpoints with filenames in the form: `epoch=01-metric=0.4367.ckpt`

The "=" sign can create issues for services where the checkpoint is uploaded, e.g. Azure/MLFlow or Kaggle Datasets since I think there are more restrictions on allowed characters. I think the issue is that if the file name ever ends up in a URL, the "=" could be confused with a query param.

Currently, I have a helper script to rename my files before upload, but it would be good if `ModelCheckpoint` could have the option to prevent this

### Pitch

Some possible solutions:

- Use an underscore instead of "="?
- Don't use a character at all, e.g. `epoch01-metric0.4367.ckpt`
- An option for the user to define the character in the kwargs?

cc @borda @awaelchli @carmocca

**Comment by awaelchli** [comment]
I don't see a way to override this right now without subclassing the ModelCheckpoint. So I think it would be reasonable to expose this symbol the same way we expose other delimeters: https://github.com/Lightning-AI/lightning/blob/6329caa876e411369e2c0876e4cec2635429d496/src/lightning/pytorch/callbacks/model_checkpoint.py#L203-L206

@carmocca thoughts?

**Comment by carmocca** [comment]
I agree :)

**Comment by subramanya1997** [comment]
I think this can be addressed by modifying this: https://github.com/Lightning-AI/lightning/blob/6329caa876e411369e2c0876e4cec2635429d496/src/lightning/pytorch/callbacks/model_checkpoint.py#L525C16-L525C39

**Comment by giorgioskij** [comment]
Hi, I would like to work on this!
It seems to me we all agree that exposing the symbol as a parameter for the user to decide is be the best option, so unless anyone disagrees, I would go that way.

**Comment by awaelchli** [comment]
Feel free to go ahead 😃 🎉

**Comment by giorgioskij** [comment]
I agree in general @carmocca , but if I'm not mistaken, since those attributes are used in `@classmethod` methods and accessed as `cls.CHECKPOINT_JOIN_CHAR`, changing the attribute on an instance should not change the value for the whole class.
Indeed, when i tested
```
checkpoint_callback = ModelCheckpoint(dirpath=".", monitor="valid_loss")
checkpoint_callback.CHECKPOINT_JOIN_CHAR = "+"
trainer = L.Trainer(callbacks=[checkpoint_callback], max_epochs=3)
```
the new symbol I set for CHECKPOINT_JOIN_CHAR was ignored, and checkpoints were saved with a "-" as per usual.

Unless I'm mistaken, which is obviously possible, I think that this structure of attributes and methods forces us to operate on the class directly. I found it a bit awkward as well.

**Comment by carmocca** [comment]
You are right in your analysis, however, I would instead make `_format_checkpoint_name` a regular method to avoid this issue

**Comment by giorgioskij** [comment]
I agree that changing the code to match the docs make more sense than the other way around in this case.
I think however that removing `@classmethod` might not be enough and we may need to rethink a bit of the structure of the class. I'm absolutely up to do it, but I would need to research a bit on class methods and attributes before jumping into an implementation.
At this point, I think it might be best to keep this issue and  PR related only to the first change (the introduction of a new attribute to change the "=" symbol), keeping the docs unchanged in the meanwhile, and create a new PR for the second topic (changing the behavior of those class attributes to make them modifiable from instances of the class and not from the class itself).
Do you agree? Thanks again for taking the time on this issue

**Comment by carmocca** [comment]
Yes. That's a great idea.

I think the `classmethod` annotation was used just to facilitate testing. It shouldn't be difficult to change

**Comment by giorgioskij** [comment]
Alright @carmocca, I just updated the PR reverting the docstring back to its previous state, like we decided. I will now set it as ready for review, and I'll make another one soon to change the code so that it matches the behavior described in the docstring.
It's a great pleasure being able to contribute to PL!

**Comment by awaelchli** [comment]
Implemented in https://github.com/Lightning-AI/lightning/pull/17999
Thanks for the contribution @giorgioskij 🎉
